app/api/ada/daedalusTransfer.js

```diff
@@ -99,6 +121,95 @@ export async function generateTransferTx(payload: {
   }
 }
 
+/**
+ * Unpack and decode provided transaction and validate all addresses and witnesses
+ * @return array of error descriptors, or string 'OK' is everything is ok
+ */
+async function _validateAddressesAndSignatures(
+  secretWords: string,
+  inputWrappers: Array<DaedalusInputWrapper>,
+  tx: MoveResponse
+): Promise<TxValidation> {
+  // Validate address/witness crypto
+  const witnessPubKeys = decodeRustTx(tx.cbor_encoded_tx).tx.witnesses.map(w => w.PkWitness[0]);
+
+  try {
+    const derivationScheme = 1;
+    const secret = await walletSecretFromMnemonic(secretWords, derivationScheme);
+    const pass = await xpubToHdPassphrase(secret.slice(64, 128));
+    const errors = inputWrappers.map((inputWrapper, idx) => {
+      const address = inputWrapper.address;
+      const addressing = inputWrapper.input.addressing;
+      const pubKey = witnessPubKeys[idx];
+      try {
+        return _validateAddressAndSignature(secret, pass, address, addressing, pubKey);
+      } catch (e) {
+        return { address, reason: 'Failed to perform validation!', error: stringifyError(e) };
+      }
+    }).filter(x => x);
```

Comment:
do this does something? I’m not so knowledgeable in dark magic :D

Reply:
Yeah, it's that the function that we are calling for each input might either return a result (indicating an error), or return nothing (`undefined`). So, like, if we map a list of three inputs and they are all ok - then we get a list like: `[undefined, undefined, undefined]`, so this filter ensures that we throw away those undefines :)

Other way to do this could be to returns something like an `Either` type from the validation function, so that there would be an object result for valid inputs too, and then do something like:
```
.filter(r => r.isError)
```

Either is fine, and I can change if this version is not readable, but I just thought, like, why create anything if we will just ignore it in the result anyways )

Other option is to rename `_validateAddressAndSignature` into something like `_getAddressSignatureError`, so it will be more logical that it might return **nothing**, when there's no error.

Reply:
thanks for the explanation!

IMO I think is fine how it is for the `filter` and the `name` )